v3/newrelic/code_level_metrics.go

```diff
@@ -33,15 +34,36 @@ type CodeLocation struct {
 	FilePath string
 }
 
+//
+// CachedCodeLocation provides storage for the code location computed such that
+// the discovery of the code location is only done once; thereafter the cached
+// value is available for use.
+//
+// This type includes methods with the same names as some of the basic code location
+// reporting functions and TraceOptions. However, when called as methods of a CachedCodeLocation value
+// instead of a stand-alone function, the operation will make use of the cache to
+// prevent computing the same source location more than once.
+//
+type CachedCodeLocation struct {
+	Location *CodeLocation
+	Err      error
+	once     sync.Once
```

Comment:
A `sync.Once` cannot be copied after first use. Unless you use a pointer here, then if someone copies a `CachedCodeLocation` around it will cause issues.

Also, I think `Location` and `Err` need to be private. There is no way for package clients to use them safely, nor is there a reason for them to do so.

Reply:
Making the `once` value a pointer shouldn't be a problem. I also agree with the suggestion to make the members private here. Really the whole `CachecCodeLocation` value should be opaque to the caller.

Reply:
Ok, I'm experimenting now with a new `CachedCodeLocation` type which uses a pointer to the `once` field so copying the struct should just give you another struct that tracks to the same `once` value instead of copying it, and has the fields all unexported. That does mean you need to call `NewCachedCodeLocation()` to get one with the pointer set up for you instead of just making a zero-value `CachedCodeLocation` so that the pointer is set up before you start doing anything concurrent in there.

Reply:
The updated code just pushed implements this approach.